internal/cache/softreservations.go

```diff
@@ -0,0 +1,172 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/sparkscheduler/v1beta1"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	v1 "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientcache "k8s.io/client-go/tools/cache"
+)
+
+// TODO(rkaram): Move to common place to avoid duplication without causing circular dependency
+const (
+	// SparkRoleLabel represents the label key for the spark-role of a pod
+	SparkRoleLabel = "spark-role"
+	// SparkAppIDLabel represents the label key for the spark application ID on a pod
+	SparkAppIDLabel = "spark-app-id" // TODO(onursatici): change this to a spark specific label when spark has one
+	// Driver represents the label key for a pod that identifies the pod as a spark driver
+	Driver = "driver"
+	// Executor represents the label key for a pod that identifies the pod as a spark executor
+	Executor = "executor"
+)
+
+// SoftReservationStore is an in-memory store that keeps track of soft reservations granted to extra executors for applications that support dynamic allocation
+type SoftReservationStore struct {
+	store     map[string]*SoftReservation // SparkAppID -> SoftReservation
+	storeLock sync.RWMutex
+}
+
+// SoftReservation is an in-memory reservation for a particular spark application that keeps track of extra executors allocate over the
+// min reservation count
+// TODO(rkaram): check if we want to use the same reservation object we already have
+type SoftReservation struct {
+	Reservations map[string]v1beta1.Reservation // Executor pod name -> Reservation (only valid ones here)
+	Status       map[string]bool                // Executor pod name -> Reservation valid or not
+}
+
+// NewSoftReservationStore builds and returns a SoftReservationStore and instantiates the needed background informer event handlers to keep the store up to date
+func NewSoftReservationStore(informer coreinformers.PodInformer) *SoftReservationStore {
+	s := &SoftReservationStore{
+		store:     make(map[string]*SoftReservation),
+		storeLock: sync.RWMutex{},
+	}
+
+	informer.Informer().AddEventHandler(
+		clientcache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				if pod, ok := obj.(*v1.Pod); ok {
+					if _, labelFound := pod.Labels[SparkRoleLabel]; labelFound {
+						return true
+					}
+				}
+				return false
+			},
+			Handler: clientcache.ResourceEventHandlerFuncs{
+				DeleteFunc: s.onPodDeletion,
+			},
+		},
+	)
+	return s
+}
+
+// GetSoftReservation returns a copy of the SoftReservation tied to an application if it exists (otherwise, bool returned will be false)
+func (s *SoftReservationStore) GetSoftReservation(appID string) (SoftReservation, bool) {
```

Comment:
Given that the other methods in this store are now exposing copies only, I think it would be better to keep retuning a copy here. What do you think?
PS: Will push an update that properly returns a deep copy of the struct.